### PR TITLE
Make device_buffer streams explicit and enforce move construction

### PR DESCRIPTION
**Pull request by harrism**
In preparation for rapidsai/rmm#775, this PR has many changes to ensure streams on which device_buffers are allocated and copied are explicit. As a consequence, this PR also finds and fixes many places where buffers were copied that should have been moved.  In addition, it adds copy constructors that take an optional stream to all `cudf::scalar` classes.

This can and should be merged before rapidsai/rmm#775 to ensure no breakage of the build by that PR.

### CODE REVIEW

File: cpp/src/lists/lists_column_factories.cu [hunk 1]
```diff
@@ -45,7 +45,7 @@ std::unique_ptr<column> make_lists_column(size_type num_rows,
   return std::make_unique<column>(cudf::data_type{type_id::LIST},
                                   num_rows,
                                   rmm::device_buffer{0, stream, mr},
-                                  null_mask,
+                                  std::forward<rmm::device_buffer>(null_mask),
```
Comment (jrhemstad) [inline cpp/src/lists/lists_column_factories.cu:48]: You only need `std::forward` when you're working with a forwarding reference (e.g., `template <typename T> (T&&)`). Here, it's just an r-value ref, so you can just use move. 
```suggestion
                                  std::move<rmm::device_buffer>(null_mask),
```

File: cpp/src/strings/strings_column_factories.cu [hunk 2]
```diff
@@ -132,7 +132,7 @@ std::unique_ptr<column> make_strings_column(size_type num_strings,
   return std::make_unique<column>(data_type{type_id::STRING},
                                   num_strings,
```
Comment (jrhemstad) [inline cpp/src/strings/strings_column_factories.cu:133]: same here

**Review by rgsl888prabhu — COMMENTED** [review]
Looks good, just comments needs to be addressed.

File: cpp/src/lists/lists_column_factories.cu (hunk 1 above)
Comment (harrism) [reply cpp/src/lists/lists_column_factories.cu]: Thanks. Fixed.

**Review by rgsl888prabhu — APPROVED** [review]

File: cpp/src/strings/strings_column_factories.cu (hunk 2 above)
Comment (harrism) [reply cpp/src/strings/strings_column_factories.cu:133]: Done.

File: cpp/src/structs/structs_column_factories.cu [hunk 3]
```diff
@@ -54,7 +54,7 @@ std::unique_ptr<cudf::column> make_structs_column(
     cudf::data_type{type_id::STRUCT},
     num_rows,
     rmm::device_buffer{0, stream, mr},  // Empty data buffer. Structs hold no data.
-    null_mask,
```
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/src/structs/structs_column_factories.cu:57]: Done.

**Review by devavret — APPROVED** [review]

**Review by jlowe — APPROVED** [review]
Java approval

File: cpp/benchmarks/type_dispatcher/type_dispatcher_benchmark.cu [hunk 4]
```diff
@@ -16,6 +16,7 @@
 
 #include "../fixture/benchmark_fixture.hpp"
 #include "../synchronization/synchronization.hpp"
+#include "rmm/device_buffer.hpp"
```
Comment (ttnghia) [inline cpp/benchmarks/type_dispatcher/type_dispatcher_benchmark.cu:19]:
```suggestion
#include <rmm/device_buffer.hpp>
```

File: cpp/include/cudf/scalar/scalar.hpp [hunk 5]
```diff
@@ -45,12 +45,23 @@ namespace cudf {
...
+
   scalar& operator=(scalar const& other) = delete;
   scalar& operator=(scalar&& other) = delete;
 
+  /**
+   * @brief Construct a new scalar object by deep copying another.
+   *
+   * @param[in] other The scalar to copy.
+   * @param[in] stream CUDA stream used for device memory operations.
+   * @param[in] mr Device memory resource to use for device memory allocation
```
Comment (ttnghia) [inline cpp/include/cudf/scalar/scalar.hpp:59]: Without having `[out]` parameter, we don't need to specify `[in]`.

File: cpp/src/dictionary/dictionary_factories.cu [hunk 6]
```diff
@@ -133,8 +133,11 @@ std::unique_ptr<column> make_dictionary_column(std::unique_ptr<column> keys,
   auto indices_column = [&] {
     // If the types match, then just commandeer the column's data buffer.
     if (new_type.id() == indices_type) {
-      return std::make_unique<column>(
-        new_type, indices_size, *(contents.data.release()), rmm::device_buffer{0, stream, mr}, 0);
+      return std::make_unique<column>(new_type,
+                                      indices_size,
+                                      std::move(*(contents.data.release())),
+                                      rmm::device_buffer{0, stream, mr},
```
Comment (ttnghia) [inline cpp/src/dictionary/dictionary_factories.cu:139]:
```suggestion
                                      rmm::device_buffer{},
```

File: cpp/src/copying/scatter.cu [hunk 7]
```diff
@@ -197,7 +197,7 @@ struct column_scalar_scatterer_impl<dictionary32, MapIterator> {
     auto contents           = new_indices->release();
     auto indices_column     = std::make_unique<column>(indices_type,
                                                    static_cast<size_type>(output_size),
-                                                   *(contents.data.release()),
+                                                   std::move(*(contents.data.release())),
                                                    rmm::device_buffer{0, stream, mr},
```
Comment (ttnghia) [inline cpp/src/copying/scatter.cu:201]:
```suggestion
                                                   rmm::device_buffer{},
```

File: cpp/include/cudf/strings/detail/copy_if_else.cuh [hunk 8]
```diff
@@ -74,8 +74,8 @@ std::unique_ptr<cudf::column> copy_if_else(
     stream,
     mr);
   size_type null_count = valid_mask.second;
-  rmm::device_buffer null_mask{0, stream, mr};
-  if (null_count) null_mask = valid_mask.first;
+  auto null_mask =
+    (null_count > 0) ? std::move(valid_mask.first) : rmm::device_buffer{0, stream, mr};
```
Comment (ttnghia) [inline cpp/include/cudf/strings/detail/copy_if_else.cuh:78]:
```suggestion
    (null_count > 0) ? std::move(valid_mask.first) : rmm::device_buffer{};
```

File: cpp/src/lists/lists_column_factories.cu [hunk 9]
```diff
@@ -45,7 +45,7 @@ std::unique_ptr<column> make_lists_column(size_type num_rows,
   return std::make_unique<column>(cudf::data_type{type_id::LIST},
                                   num_rows,
                                   rmm::device_buffer{0, stream, mr},
```
Comment (ttnghia) [inline cpp/src/lists/lists_column_factories.cu:47]:
```suggestion
                                  rmm::device_buffer{},
```

File: cpp/src/replace/clamp.cu [hunk 10]
```diff
@@ -324,7 +324,7 @@ std::unique_ptr<column> dispatch_clamp::operator()<cudf::dictionary32>(
   auto contents           = new_indices->release();
   auto indices_column     = std::make_unique<column>(indices_type,
                                                  static_cast<size_type>(output_size),
-                                                 *(contents.data.release()),
+                                                 std::move(*(contents.data.release())),
                                                  rmm::device_buffer{0, stream, mr},
```
Comment (ttnghia) [inline cpp/src/replace/clamp.cu:328]:
```suggestion
                                                 rmm::device_buffer{},
```

File: cpp/src/replace/replace.cu [hunk 11]
```diff
@@ -469,8 +469,11 @@ std::unique_ptr<cudf::column> replace_kernel_forwarder::operator()<cudf::diction
     mr);
   auto null_count     = new_indices->null_count();
   auto contents       = new_indices->release();
-  auto indices_column = std::make_unique<cudf::column>(
-    indices_type, input.size(), *(contents.data.release()), rmm::device_buffer{0, stream, mr}, 0);
+  auto indices_column = std::make_unique<cudf::column>(indices_type,
+                                                       input.size(),
+                                                       std::move(*(contents.data.release())),
+                                                       rmm::device_buffer{0, stream, mr},
```
Comment (ttnghia) [inline cpp/src/replace/replace.cu:475]:
```suggestion
                                                       rmm::device_buffer{},
```

File: cpp/src/strings/strings_column_factories.cu [hunk 12]
```diff
@@ -132,7 +132,7 @@ std::unique_ptr<column> make_strings_column(size_type num_strings,
...
                                   rmm::device_buffer{0, stream, mr},
```
Comment (ttnghia) [inline cpp/src/strings/strings_column_factories.cu:134]:
```suggestion
                                  rmm::device_buffer{},
```

File: cpp/src/structs/structs_column_factories.cu [hunk 13]
```diff
@@ -54,7 +54,7 @@ std::unique_ptr<cudf::column> make_structs_column(
     cudf::data_type{type_id::STRUCT},
     num_rows,
     rmm::device_buffer{0, stream, mr},  // Empty data buffer. Structs hold no data.
```
Comment (ttnghia) [inline cpp/src/structs/structs_column_factories.cu:56]:
```suggestion
    rmm::device_buffer{},  // Empty data buffer. Structs hold no data.
```

**Review by ttnghia — APPROVED** [review]

File: cpp/include/cudf/scalar/scalar.hpp (hunk 5 above)
Comment (harrism) [reply cpp/include/cudf/scalar/scalar.hpp:59]: I did not add the [in].

**Comment by harrism** [comment]
@gpucibot merge

**Comment by ttnghia** [comment]
Rerun tests.

**Comment by ttnghia** [comment]
Rerun tests.

**Pull request merged**